content/en/blog/_posts/2019-10-15-csi-driver-testing.md

```diff
@@ -0,0 +1,317 @@
+specified with `-csi.controllerendpoint`. Directories are created in
+`/tmp` by default. This can be changed via `-csi.mountdir` and
+`-csi.stagingdir`.
+
+Some drivers cannot be deployed such that everything is guaranteed to
+run on the same host. In such a case, custom scripts have to be used
+to handle directories: they log into the host where the CSI node
+controller runs and create or remove the directories there.
+
+For example, the [PMEM-CSI](https://github.com/intel/pmem-csi)
+driver manages persistent memory, a local resource that is often not
+available on typical development machines. It gets tested by deploying
+in a Kubernetes cluster that runs inside QEMU virtual machines with a
+deployment .yaml file that explicitly enables connecting to the driver
+from remote by forwarding incoming TCP connections with
+[`socat`]((https://github.com/intel/pmem-csi/blob/82e6114a7282015cabc0fdcd46c24776bf3876dd/deploy/kubernetes-1.14/pmem-csi-direct-testing.yaml#L221-L229).
+
+The Kubernetes-CSI example driver also gets tested on Kubernetes this
+way. Here's how one can replicate that, using
+[`csi-driver-host-path`](https://github.com/kubernetes-csi/csi-driver-host-path) v1.2.0:
+
+```
+$ cd csi-driver-host-path
+$ git describe --tags HEAD
+v1.2.0
+$ kubectl get nodes
+NAME        STATUS   ROLES    AGE   VERSION
+127.0.0.1   Ready    <none>   42m   v1.16.0
+
+$ cat >mkdir_in_pod.sh <<EOF
+#!/bin/sh
+kubectl exec csi-hostpathplugin-0 -c hostpath -- mktemp -d /tmp/csi-sanity.XXXXXX
+EOF
+
+$ cat >rmdir_in_pod.sh <<EOF
+#!/bin/sh
+kubectl exec csi-hostpathplugin-0 -c hostpath -- rmdir "\$@"
+EOF
+
+$ chmod u+x *_in_pod.sh
+$ csi-sanity -ginkgo.v \
+             -csi.junitfile junit_sanity.xml \
+             -csi.endpoint dns:///127.0.0.1:$(kubectl get "services/hostpath-service" -o "jsonpath={..nodePort}") \
+             -csi.createstagingpathcmd ./mkdir_in_pod.sh \
+             -csi.createmountpathcmd ./mkdir_in_pod.sh \
+             -csi.removestagingpathcmd ./rmdir_in_pod.sh \
+             -csi.removemountpathcmd ./rmdir_in_pod.sh
+
+Running Suite: CSI Driver Test Suite
+====================================
+Random Seed: 1570540138
+Will run 72 of 72 specs
+...
+Controller Service [Controller Server] ControllerGetCapabilities 
+  should return appropriate capabilities
+  /nvme/gopath/src/github.com/kubernetes-csi/csi-test/pkg/sanity/controller.go:111
+STEP: connecting to CSI driver
+STEP: creating mount and staging directories
+STEP: checking successful response
+•
+------------------------------
+Controller Service [Controller Server] GetCapacity 
+  should return capacity (no optional values added)
+  /nvme/gopath/src/github.com/kubernetes-csi/csi-test/pkg/sanity/controller.go:149
+STEP: reusing connection to CSI driver at dns:///127.0.0.1:30056
+STEP: creating mount and staging directories
+...
+Ran 53 of 72 Specs in 148.206 seconds
+SUCCESS! -- 53 Passed | 0 Failed | 0 Pending | 19 Skipped
+PASS
+```
+
```

Comment:
Can we link (or just refer) to the implementation of the hostpath sanity tests for readers to have something to model after?

Reply:
These are the tests from https://github.com/kubernetes-csi/csi-test/tree/master/pkg/sanity, i.e. not some tests that are specific to hostpath.

I added:
* The source code of these tests is in the [`pkg/sanity`
  package](https://github.com/kubernetes-csi/csi-test/tree/master/pkg/sanity).